controlplane/pkg/apis/nsm/connection/connection.go

```diff
@@ -0,0 +1,44 @@
+package connection
+
+import "github.com/networkservicemesh/networkservicemesh/controlplane/pkg/apis/connectioncontext"
+
+// State is a enum for connection state
+type State int8
+
+const (
+	// StateUp means that connection is up
+	StateUp State = 0
+	// StateDown means that connection is down
+	StateDown State = 1
+)
+
+// Connection is an unified interface for local/remote connections
+type Connection interface {
+	IsRemote() bool
```

Comment:
consider about extracting interfaces from this interface

Reply:
I am not actually sure if there is a case when we can use such interface:
```
type ApiObject interface {
    IsRemote() bool
    Equals(other ApiObject)
    Clone() ApiObject
}
```
Can you please share your ideas about this?

Reply:
As you can see methods `IsRemote()`, `IsValid() error`, `IsComplete() error` are used in couple of places. The main idea considers creating share interfaces for these interfaces.
Something like:
```
type ShareInterace interface{
    IsRemote() bool
    IsValid() error
}
```
```
type Connection interface {
    ShareInterface
    ...
}
```
```
type Mechanism interface{
    ShareInterface
    ...
}
```

Reply:
If we use such methods like `IsRemote()`, `IsValid()` we mostly want to distinguish the type of the passed value if it is `Connection`, or `Request`, or something other.

On the other hand, there are only few places where we use those mentioned methods. For the rest of the project we need to use only getter and setter methods.
So if there is a point to narrow interfaces, I would rather prefer to do something like:
```
type Shared interface {
    IsRemote() bool
    IsValid() error
}

type Connection interface {
    Equals(connection Connection) bool
    Clone() Connection

    // getter and setter methods
}

type ExtendedConnection interface {
    Shared
    Connection
}
```

But for me it looks like overcomplicating the code instead of making it better.